### PR TITLE
:art: remove data object on simpleObject store

### PR DESCRIPTION
**Pull request by guillaumecrespel**

### CODE REVIEW

File: src/selectors.spec.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ const Element = code => ({ code, some: 'other', infos: code })
 const SubElement = code => subCode => ({ ...Element(code), sub: { subCode } })
 
 const path = 'api.raw'
-const prefix = 'testPrefix'
+const name = 'testPrefix'
```
Comment (fabienjuif) [inline src/selectors.spec.js]: testName ?

File: src/types/keyValue/index.js [hunk 2]
```diff
@@ -3,5 +3,5 @@ import middleware from './keyValue.middleware'
 export default {
   middlewares: [middleware],
   actions: ['set', 'add', 'reset', 'remove', 'update'],
-  selectors: ['getState', 'getKeys', 'getAsArray', 'getLength', 'isInitialized', 'get', 'getBy'],
+  selectorsEnabled: ['getState', 'getKeys', 'getAsArray', 'getLength', 'isInitialized', 'get', 'getBy'],
```
Comment (fabienjuif) [inline src/types/keyValue/index.js]: export selected selectors from common to a `keyValue.selectors.js'
then 
```es6
import * as selectors from './keyValue.selectors'

export  default {
  // ...
  selectors,
}
```

File: src/selectors.js [hunk 3]
```diff
@@ -1,21 +1,21 @@
...
 export const getKeys = getFactory('keys')
 export const getAsArray = getFactory('array')
-export const getLength = path => prefix => state => getKeys(path)(prefix)(state).length
+export const getLength = ({ path, name }) => state => getKeys({ path, name })(state).length
 export const isInitialized = getFactory('initialized')
 
-const getData = getFactory('data')
-export const get = path => prefix => keys => (state) => {
-  const data = getData(path)(prefix)(state)
+export const getData = options => getFactory('data')(options)
```
Comment (fabienjuif) [inline src/selectors.js]: Why do you export getData ?

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
+/* eslint-env jest */
+import { isInitialized, get } from './simpleObject.selectors'
+import { initState } from './simpleObject.middleware'
+
+describe('selectors/simpleObject => ', () => {
+  describe('isInitialized() :', () => {
+    describe('without defaultData', () => {
+      const withoutDefault = { name: 'o' }
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: const options

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      it('initialized with object value', () => {
+        expect(isInitialized(withoutDefault)({ o: { i: 'initialized' } })).toMatchSnapshot()
+      })
+      it('initialized with string value', () => {
+        expect(isInitialized(withoutDefault)({ o: 'initialized' })).toMatchSnapshot()
+      })
+    })
+
+    describe('defaultData should be a string', () => {
+      const string = { name: 'o', defaultData: 'defaultData' }
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: const options

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+      it('not initialized', () => {
+        expect(isInitialized(string)({ o: string.defaultData })).toMatchSnapshot()
+      })
+      it('initialized', () => {
+        expect(isInitialized(string)({ o: 'initialized' })).toMatchSnapshot()
+      })
+    })
+
+    describe('defaultData should be an object', () => {
+      const object = { name: 'o', defaultData: {} }
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: const options

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  describe('get(): ', () => {
+    describe('without defaultData', () => {
+      const withoutDefault = { name: 'o' }
+      it('and get the initState', () => {
+        expect(get(withoutDefault)()({ o: initState })).toMatchSnapshot()
+      })
+    })
+
+    describe('defaultData should be a string', () => {
+      const string = { name: 'o', defaultData: '' }
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: const options

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+      it('and get the default data', () => {
+        expect(get(string)()({ o: string.defaultData })).toMatchSnapshot()
+      })
+      it('and get the data', () => {
+        expect(get(string)()({ o: 'DATA' })).toMatchSnapshot()
+      })
+    })
+
+    describe('defaultData should be an object', () => {
+      const object = { name: 'o', defaultData: {} }
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: const options

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
+/* eslint-env jest */
+import { isInitialized, get } from './simpleObject.selectors'
+import { initState } from './simpleObject.middleware'
+
+describe('selectors/simpleObject => ', () => {
+  describe('isInitialized() :', () => {
+    describe('without defaultData', () => {
+      const withoutDefault = { name: 'o' }
+      it('not initialized', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should not be initialized

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        expect(isInitialized(withoutDefault)({ o: initState })).toMatchSnapshot()
+      })
+      it('initialized with object value', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should be initialized -object-

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        expect(isInitialized(withoutDefault)({ o: { i: 'initialized' } })).toMatchSnapshot()
+      })
+      it('initialized with string value', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should be initialized -string-

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        expect(isInitialized(withoutDefault)({ o: 'initialized' })).toMatchSnapshot()
+      })
+    })
+
+    describe('defaultData should be a string', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: with defaultData as a string

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+      const string = { name: 'o', defaultData: 'defaultData' }
+      it('not initialized', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should not be initialized

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        expect(isInitialized(string)({ o: string.defaultData })).toMatchSnapshot()
+      })
+      it('initialized', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should be initialized

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+        expect(isInitialized(string)({ o: 'initialized' })).toMatchSnapshot()
+      })
+    })
+
+    describe('defaultData should be an object', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: with defaultData as an object

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+      const object = { name: 'o', defaultData: {} }
+      it('not initialized', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should not be initialized

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+        expect(isInitialized(object)({ o: object.defaultData })).toMatchSnapshot()
+      })
+      it('initialized', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should be initialized

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+        expect(isInitialized(object)({ o: { i: 'initialized' } })).toMatchSnapshot()
+      })
+    })
+  })
+
+  describe('get(): ', () => {
+    describe('without defaultData', () => {
+      const withoutDefault = { name: 'o' }
+      it('and get the initState', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should returns de initState

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+        expect(get(withoutDefault)()({ o: initState })).toMatchSnapshot()
+      })
+    })
+
+    describe('defaultData should be a string', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: defaultData as a string

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+      const string = { name: 'o', defaultData: '' }
+      it('and get the default data', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should return the default data

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+        expect(get(string)()({ o: string.defaultData })).toMatchSnapshot()
+      })
+      it('and get the data', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should return the current data

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+        expect(get(string)()({ o: 'DATA' })).toMatchSnapshot()
+      })
+    })
+
+    describe('defaultData should be an object', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: defaultData as an object

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+      const object = { name: 'o', defaultData: {} }
+      it('and get the default data', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should return the default data

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+        expect(get({ name: 'o' })()({ o: object.defaultData })).toMatchSnapshot()
+      })
+      it('and get the data', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: should return the current data

File: src/types/simpleObject/simpleObject.selectors.js [hunk 25]
```diff
@@ -0,0 +1,10 @@
+/* eslint-disable import/prefer-default-export */
+import { initState } from './simpleObject.middleware'
+import { getState } from '../../selectors'
+
+export const get = options => () => state => getState(options)(state)
+
+export const isInitialized = options => (state) => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.js]: use getState here

File: src/index.spec.js [hunk 26]
```diff
@@ -113,6 +113,19 @@ describe('index', () => {
     })
   })
 
+  describe('extends selectors', () => {
```
Comment (fabienjuif) [inline src/index.spec.js]: extended selectors

Comment (guillaumecrespel) [inline src/index.spec.js]: :ok_hand:

File: src/types/simpleObject/simpleObject.selectors.js (hunk 25 above)
Comment (guillaumecrespel) [inline src/types/simpleObject/simpleObject.selectors.js]: :call_me_hand:

**Comment by fabienjuif** [comment]
About TYPES.md to remove `getState` from API documentation

File: src/__snapshots__/index.spec.js.snap [hunk 27]
```diff
@@ -2,6 +2,10 @@
 
 exports[`index action type factories should export action type factories 1`] = `"@trampss/SET_A SIMPLEOBJECT"`;
 
+exports[`index extended selectors should extends get() selector on simpleObject 1`] = `"DATA"`;
```
Comment (fabienjuif) [inline src/__snapshots__/index.spec.js.snap]: See my comment about `()`

File: src/index.spec.js [hunk 28]
```diff
@@ -113,9 +113,19 @@ describe('index', () => {
     })
   })
 
+  describe('extended selectors', () => {
+    const so = simpleObject({ name: 'o', defaultData: {} })
```
Comment (fabienjuif) [inline src/index.spec.js]: so is not reused

File: src/index.spec.js [hunk 29]
```diff
@@ -113,9 +113,19 @@ describe('index', () => {
...
+    })
+    it('should extends get() selector on simpleObject', () => {
+      expect(so.get()({ o: 'DATA' })).toMatchSnapshot()
+    })
+  })
+
   describe('action type factories', () => {
     it('should export action type factories', () => {
-      expect(simpleObject({ name: 'a simpleObject' }).SET).toMatchSnapshot()
+      expect(simpleObject({ name: 'a simpleObject' }).set().type).toMatchSnapshot()
```
Comment (fabienjuif) [inline src/index.spec.js]: If you are doing this, this is a major version change

File: src/selectors.js [hunk 30]
```diff
@@ -1,21 +1,21 @@
...
-  return subState[prefix]
+  return subState[options.name]
 }
 
-const getFactory = key => path => prefix => state => getState(path)(prefix)(state)[key]
+const getFactory = key => options => state => getState(options)(state)[key]
 export const getKeys = getFactory('keys')
 export const getAsArray = getFactory('array')
-export const getLength = path => prefix => state => getKeys(path)(prefix)(state).length
+export const getLength = ({ path, name }) => state => getKeys({ path, name })(state).length
```
Comment (fabienjuif) [inline src/selectors.js]: `export const getLength = (options) => state => getKeys(options)(state).length
`

Comment (fabienjuif) [inline src/selectors.js]: same for `get` and `getBy`

File: src/types/simpleObject/simpleObject.selectors.js [hunk 31]
```diff
@@ -0,0 +1,11 @@
...
+import { getState } from '../../selectors'
+
+const get = options => () => state => getState(options)(state)
+
+const isInitialized = options => (state) => {
+  if (options.defaultData !== undefined) return getState(options)(state) !== options.defaultData
+  return get(options)()(state) !== initState
+}
+
+export { get, isInitialized }
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.js]: why don't you export in a onliner ?
```

export const get
export const isInitialized

```
As we do everywhere

**Comment by fabienjuif** [comment]
Don't forget about : https://github.com/Trampss/trampss-redux-factory/pull/59#issuecomment-318317337

File: src/__snapshots__/index.spec.js.snap (hunk 27 above)
Comment (fabienjuif) [inline src/__snapshots__/index.spec.js.snap]: up

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+      it('should not be initialized', () => {
+        expect(selectors.isInitialized(options)({ o: options.defaultData })).toMatchSnapshot()
+      })
+      it('should be initialized', () => {
+        expect(selectors.isInitialized(options)({ o: { i: 'initialized' } })).toMatchSnapshot()
+      })
+    })
+  })
+
+  describe('get: ', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: remove extra space

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 33]
```diff
@@ -0,0 +1,69 @@
+/* eslint-env jest */
+import * as selectors from './simpleObject.selectors'
+import { initState } from './simpleObject.middleware'
+
+describe('selectors/simpleObject => ', () => {
+  describe('isInitialized :', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: remove space before :

File: src/index.spec.js [hunk 34]
```diff
@@ -113,6 +113,15 @@ describe('index', () => {
     })
   })
 
+  describe('extended selectors', () => {
+    it('should extends isInitialized() selector on simpleObject', () => {
+      expect(simpleObject({ name: 'o' }).isInitialized({ o: '' })).toMatchSnapshot()
+    })
+    it('should extends get() selector on simpleObject', () => {
```
Comment (fabienjuif) [inline src/index.spec.js]: ()

File: src/index.spec.js [hunk 35]
```diff
@@ -113,6 +113,15 @@ describe('index', () => {
     })
   })
 
+  describe('extended selectors', () => {
+    it('should extends isInitialized() selector on simpleObject', () => {
```
Comment (fabienjuif) [inline src/index.spec.js]: ()

File: src/selectors.js [hunk 36]
```diff
@@ -1,21 +1,21 @@
...
 export const getAsArray = getFactory('array')
-export const getLength = path => prefix => state => getKeys(path)(prefix)(state).length
+export const getLength = options => state => getKeys(options)(state).length
 export const isInitialized = getFactory('initialized')
 
-const getData = getFactory('data')
-export const get = path => prefix => keys => (state) => {
-  const data = getData(path)(prefix)(state)
+const getData = options => getFactory('data')(options)
+export const get = ({ path, name }) => keys => (state) => {
```
Comment (fabienjuif) [inline src/selectors.js]: get = (options)

File: src/selectors.js [hunk 37]
```diff
@@ -1,21 +1,21 @@
...
+  const data = getData({ path, name })(state)
```
Comment (fabienjuif) [inline src/selectors.js]: getData(options)

File: src/selectors.js [hunk 38]
```diff
@@ -24,8 +24,8 @@ export const get = path => prefix => keys => (state) => {
   return data[keys]
 }
 
-export const getBy = path => prefix => (propertyPath, values) => (state) => {
-  const data = getAsArray(path)(prefix)(state)
+export const getBy = ({ path, name }) => (propertyPath, values) => (state) => {
```
Comment (fabienjuif) [inline src/selectors.js]: getBy = (options)

File: src/selectors.js [hunk 39]
```diff
@@ -24,8 +24,8 @@ export const get = path => prefix => keys => (state) => {
...
+  const data = getAsArray({ path, name })(state)
```
Comment (fabienjuif) [inline src/selectors.js]: getAsArray(options)

File: README.md [hunk 40]
```diff
@@ -82,7 +82,7 @@ Parameters are :
      - it's used to retrieve informations from selectors
    - **prefix** (optional) is added to actions to avoid some collisions when there are two reducers with the same name in two distincts paths
    - **type** (optional) can be `keyValue` or `simpleObject` (default is `keyValue`)
-   - **defaultData** (optional), for `simpleObject` only, set the default data value (used by `reset` action and for initialisation)
+   - **defaultData** (optional), for `simpleObject` only, set the default data value (used by `reset` action and for initialisation). If it not used, the default value is an empty object `{}`
```
Comment (fabienjuif) [inline README.md]: `(default is an empty object `{}`)`

Comment (guillaumecrespel) [inline README.md]: :+1:

File: package.json [hunk 41]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "trampss-redux-factory",
-  "version": "3.2.0",
+  "version": "3.2.1",
```
Comment (fabienjuif) [inline package.json]: what about 3.3.0 since tests are broken ?

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 42]
```diff
@@ -0,0 +1,69 @@
...
+      })
+      it('should be initialized -object-', () => {
+        expect(selectors.isInitialized(options)({ o: { i: 'initialized' } })).toMatchSnapshot()
+      })
+      it('should be initialized -string-', () => {
+        expect(selectors.isInitialized(options)({ o: 'initialized' })).toMatchSnapshot()
+      })
+    })
+
+    describe('defaultData should be a string', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: defaultData as a string

File: src/types/simpleObject/simpleObject.selectors.spec.js [hunk 43]
```diff
@@ -0,0 +1,69 @@
...
+      const options = { name: 'o', defaultData: 'defaultData' }
+      it('should not be initialized', () => {
+        expect(selectors.isInitialized(options)({ o: options.defaultData })).toMatchSnapshot()
+      })
+      it('should be initialized', () => {
+        expect(selectors.isInitialized(options)({ o: 'initialized' })).toMatchSnapshot()
+      })
+    })
+
+    describe('defaultData should be an object', () => {
```
Comment (fabienjuif) [inline src/types/simpleObject/simpleObject.selectors.spec.js]: defaultData as an object

**Pull request merged**